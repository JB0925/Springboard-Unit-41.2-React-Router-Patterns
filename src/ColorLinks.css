.ColorLinks {
    text-align: center;
}

ul li {
    list-style: none;
}

ul a {
    font-size: 1.8rem;
    color: black;
}