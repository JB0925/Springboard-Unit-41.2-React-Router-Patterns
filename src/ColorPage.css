.ColorPage {
    color: white;
}

.ColorPage h1, a {
    font-size: 5rem;
    color: white;
}