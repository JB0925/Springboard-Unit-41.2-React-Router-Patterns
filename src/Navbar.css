nav {
    display: flex;
    align-items: center;
    justify-content: space-around;
    background-color: darkslateblue;
}

nav h1 {
    color: white;
    font-size: 1.8rem;
}

nav button {
    background-color: dodgerblue;
    color: white;
    padding: 10px;
    border: 1px solid dodgerblue;
    border-radius: 5px;
    margin-left: 25px;
}

label {
    color: white;
}

input {
    margin-left: 25px;
    padding: 7px;
    border-radius: 5px;
}

