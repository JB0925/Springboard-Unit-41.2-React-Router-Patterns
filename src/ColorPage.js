import React from "react";
import { useParams, Link } from "react-router-dom";
import "./ColorPage.css";

const ColorPage = () => {
    const { color } = useParams();
    const style = {
        backgroundColor: color,
        height: "100vh",
        width: "100vw"
    };

    return (
        <div className="ColorPage" style={style}>
            <h1>This page is {color}!!!</h1>
            <Link to="/colors">GO BACK</Link>
        </div>
    );
};

export default ColorPage;