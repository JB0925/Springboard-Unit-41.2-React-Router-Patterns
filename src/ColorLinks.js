import React from "react";
import { Link } from "react-router-dom";
import { useColorAdd } from "./hooks/hooks";
import "./ColorLinks.css";
import Navbar from "./Navbar";

const ColorLinks = () => {
    const [colors, addColor] = useColorAdd();
    const colorLinks = colors.map(c => <li><Link to={`/colors/${c}`}>{c}</Link></li>);

    return (
        <div className="ColorLinks">
            <Navbar addColor={addColor} />
            <h2>Please Select a Color</h2>
            <ul>
                {colorLinks}
            </ul>
        </div>
    );
};

export default ColorLinks;