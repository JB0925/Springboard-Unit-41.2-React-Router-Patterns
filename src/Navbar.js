import React from "react";
import { useSetFormState, useSubmitState } from "./hooks/hooks";
import "./Navbar.css";

const Navbar = ({ addColor }) => {
    const [formData, updateForm, submitForm] = useSetFormState();
    const [formDisplayed, toggleFormDisplay] = useSubmitState();

    const optionalRender = () => {
        if (!formDisplayed) {
            return <button onClick={toggleFormDisplay}>Add Color</button>
        };
        return (
            <form onSubmit={evt => submitForm(addColor, toggleFormDisplay, evt)}>
                <label htmlFor="color">Enter a Color</label>
                <input
                    id="color"
                    name="color"
                    placeholder="i.e. red, blue, etc."
                    onChange={updateForm}
                    value={formData.color}
                    type="text"
                    required
                />
                <button type="submit">Add Color</button>
            </form>
        );
    };

    return (
        <nav>
            <h1>Welcome to the Color Factory!</h1>
            {optionalRender()}
        </nav>
    );
};

export default Navbar;