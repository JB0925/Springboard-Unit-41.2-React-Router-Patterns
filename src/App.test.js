import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router-dom';
import App from './App';
import Navbar from "./Navbar";
import ColorLinks from "./ColorLinks";
import ColorPage from "./ColorPage";

describe("smoke tests for the color pages app", () => {
  test("the App component will render", () => {
    render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );
  });

  test("the Navbar component renders", () => {
    render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    );
  });

  test("the ColorLinks component renders", () => {
    render(
      <MemoryRouter>
        <ColorLinks />
      </MemoryRouter>
    );
  });

  test("the ColorPage component renders", () => {
    render(
      <MemoryRouter initialEntries={["/colors/blue"]}>
        <Route path="/colors/:name">
          <ColorPage />
        </Route>
      </MemoryRouter>
    );
  });
});

describe("snapshot tests for each component", () => {
  test("The App component matches the snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("The Navbar component matches the snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("The ColorLinks component matches the snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={["/colors/blue"]}>
        <Route path="/colors/:name">
          <ColorPage />
        </Route>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("submitting the form adds a new link", () => {
  test("submitting the form should render a new link", () => {
    // Render the App Component
    const { getByText, getByLabelText } = render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );

    // Check that the "Add Color" button is in the document
    const addBtn = getByText("Add Color");
    expect(addBtn).toBeInTheDocument();
    
    // Click the button and ensure that the form submission button / form is showing
    fireEvent.click(addBtn);
    const submitBtn = getByText("Add Color");
    expect(submitBtn).toBeInTheDocument();
    
    // Check that the input is in the document
    const input = getByLabelText("Enter a Color");
    expect(input).toBeInTheDocument();
    
    // Enter "purple" into the input, and submit the form
    // Finally, check to ensure that that correct HTML is rendered.
    fireEvent.change(input, { target: { value: "purple" }});
    expect(document.body).not.toContainHTML(`<li><a href="/colors/purple">purple</a></li>`);
    fireEvent.click(submitBtn);
    expect(document.body).toContainHTML(`<li><a href="/colors/purple">purple</a></li>`);
  });
});


