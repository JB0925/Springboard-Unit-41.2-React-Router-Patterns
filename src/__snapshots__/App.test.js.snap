// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests for each component The App component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="ColorLinks"
    >
      <nav>
        <h1>
          Welcome to the Color Factory!
        </h1>
        <button>
          Add Color
        </button>
      </nav>
      <h2>
        Please Select a Color
      </h2>
      <ul />
    </div>
  </div>
</DocumentFragment>
`;

exports[`snapshot tests for each component The ColorLinks component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ColorPage"
    style="height: 100vh; width: 100vw;"
  >
    <h1>
      This page is !!!
    </h1>
    <a
      href="/colors"
    >
      GO BACK
    </a>
  </div>
</DocumentFragment>
`;

exports[`snapshot tests for each component The Navbar component matches the snapshot 1`] = `
<DocumentFragment>
  <nav>
    <h1>
      Welcome to the Color Factory!
    </h1>
    <button>
      Add Color
    </button>
  </nav>
</DocumentFragment>
`;
